File: app/packages/web/scripts/sdk/targets/python/runtime.py
"""The hand-written half of the generated Python SDK.

This module is **not imported by anything in this repo** — the generator reads
it off disk and writes everything below the sentinel out as
``infrawrench_sdk/_transport.py``, under the usual generated-file banner.
Keeping it as real Python (rather than a template string inside ``index.ts``)
means it can be run, linted and type-checked like any other source file, so a
mistake in the request plumbing shows up here instead of in generated output
nobody reads.

Three tokens are substituted at generation time — see ``SUBSTITUTIONS`` in
``./index.ts``. They are written as plain literals so this file still parses,
imports and type-checks on its own.

Nothing here may import a third-party package: the published distribution has
no dependencies and must run on a stock CPython 3.9.
"""

# --8<-- everything below this line is copied into the generated SDK --8<--

from __future__ import annotations

import json
import re
import urllib.error
import urllib.parse
import urllib.request
import uuid
from dataclasses import dataclass, field
from typing import Any, Dict, List, Mapping, Optional, Tuple, Union

# Replaced with the first server advertised by the spec.
DEFAULT_BASE_URL = "@@BASE_URL@@"

# Replaced with the path parameter the client carries as configuration
# (``orgId``), and the constructor keyword that supplies it — or ``None`` when
# the API has no such parameter.
SCOPE_PARAM: Optional[str] = "@@SCOPE_PARAM@@"
SCOPE_KWARG: Optional[str] = "@@SCOPE_KWARG@@"

# What a ``format: binary`` field accepts. Bare bytes are uploaded under the
# field's own name; the tuple forms let a caller pin the filename and, with the
# three-element form, the part's content type.
FileUpload = Union[
    bytes,
    bytearray,
    Tuple[str, Union[bytes, bytearray]],
    Tuple[str, Union[bytes, bytearray], str],
]

_PLACEHOLDER = re.compile(r"\{([^}]+)\}")

_CRLF = b"\r\n"

_DEFAULT_PART_TYPE = "application/octet-stream"


@dataclass
class ClientOptions:
    """Everything :class:`APIV1Client` can be configured with.

    Passed either as keyword arguments to the client — the usual way — or as a
    prebuilt instance via ``options=``, which is handy when the configuration
    is assembled somewhere other than the call site.
    """

    #: Base URL of the deployment. Defaults to the production API.
    base_url: Optional[str] = None
    #: API key or WorkOS access token, sent as ``Authorization: Bearer <token>``.
    #: Omit it only if you are supplying the header yourself.
    api_key: Optional[str] = None
    #: Default organization id. Every org-scoped call accepts ``org_id``; set it
    #: here once and you can leave it off the call sites.
    org_id: Optional[str] = None
    #: Headers merged into every request. Per-call headers win.
    headers: Optional[Mapping[str, str]] = None
    #: Socket timeout in seconds, applied to every request. No timeout by default.
    timeout: Optional[float] = None
    #: Swap the URL opener — for proxies, custom TLS contexts, or tests.
    opener: Optional[urllib.request.OpenerDirector] = None


@dataclass
class RequestOptions:
    """Per-call overrides, accepted as ``request_options`` by every method."""

    #: Headers for this call only, merged over the client-wide ones.
    headers: Optional[Mapping[str, str]] = None
    #: Timeout in seconds for this call only, overriding the client-wide one.
    timeout: Optional[float] = None


class ApiError(Exception):
    """Raised for any non-2xx response.

    The parsed response body is on :attr:`body`, and :attr:`code` carries the
    machine-readable discriminator when the API sends one (for example
    ``reauthentication_required`` on a step-up 403) — branch on that, not on the
    message.
    """

    def __init__(
        self,
        *,
        status: int,
        message: str,
        code: Optional[str],
        body: Any,
        method: str,
        url: str,
    ) -> None:
        super().__init__(message)
        self.status = status
        self.code = code
        self.body = body
        self.method = method
        self.url = url

    def __repr__(self) -> str:
        return "ApiError(status={0!r}, code={1!r}, method={2!r}, url={3!r})".format(
            self.status, self.code, self.method, self.url
        )


@dataclass
class Response:
    """A raw HTTP response, as returned by :meth:`ApiTransport.send`."""

    status: int
    #: Header names lowercased, so lookups don't have to guess the casing.
    headers: Dict[str, str] = field(default_factory=dict)
    body: bytes = b""


class ApiTransport:
    """Request plumbing shared by every namespace.

    Public because the generated namespace classes take one in their
    constructor and because :meth:`send` is the documented seam for tests and
    proxies — but reach for :class:`APIV1Client` to make actual calls.
    """

    def __init__(self, options: Optional[ClientOptions] = None, **overrides: Any) -> None:
        config = options if options is not None else ClientOptions()
        for key, value in overrides.items():
            if not hasattr(config, key):
                raise TypeError("Unknown client option {0!r}".format(key))
            if value is not None:
                setattr(config, key, value)

        #: Normalized base URL, without a trailing slash.
        self.base_url = (config.base_url or DEFAULT_BASE_URL).rstrip("/")
        self._api_key = config.api_key
        self._defaults: Dict[str, Optional[str]] = (
            {} if SCOPE_PARAM is None else {SCOPE_PARAM: config.org_id}
        )
        self._headers = {key.lower(): value for key, value in (config.headers or {}).items()}
        self._timeout = config.timeout
        self._opener = config.opener or urllib.request.build_opener()

    def request(
        self,
        *,
        method: str,
        path: str,
        path_params: Optional[Mapping[str, Any]] = None,
        query: Optional[Mapping[str, Any]] = None,
        body: Any = None,
        form: Optional[Mapping[str, Any]] = None,
        accept: str = "json",
        options: Optional[RequestOptions] = None,
    ) -> Any:
        """Perform one call and decode its response according to ``accept``."""
        route = self._resolve_path(method, path, path_params)
        url = self.base_url + route + _encode_query(query)

        headers = {
            "accept": "application/octet-stream" if accept == "binary" else "application/json"
        }
        headers.update(self._headers)
        if options is not None and options.headers:
            headers.update({key.lower(): value for key, value in options.headers.items()})
        if self._api_key:
            headers["authorization"] = "Bearer {0}".format(self._api_key)

        data: Optional[bytes] = None
        if form is not None:
            data, content_type = _encode_multipart(form)
            headers["content-type"] = content_type
        elif body is not None:
            data = json.dumps(body).encode("utf-8")
            headers["content-type"] = "application/json"

        timeout = self._timeout
        if options is not None and options.timeout is not None:
            timeout = options.timeout

        response = self.send(method.upper(), url, headers, data, timeout)

        if not 200 <= response.status < 300:
            raise _to_api_error(response, method.upper(), url)
        if accept == "binary":
            return response.body
        if accept == "empty" or response.status in (204, 205) or not response.body:
            return None
        text = response.body.decode("utf-8", "replace")
        if "json" not in response.headers.get("content-type", ""):
            return text
        return json.loads(text)

    def send(
        self,
        method: str,
        url: str,
        headers: Mapping[str, str],
        data: Optional[bytes],
        timeout: Optional[float],
    ) -> Response:
        """Put one prepared request on the wire.

        The single point where this package touches the network, so overriding
        it — in a subclass or by assignment — is all it takes to record, replay
        or fake every call the client makes.
        """
        request = urllib.request.Request(url=url, data=data, method=method)
        for key, value in headers.items():
            request.add_header(key, value)
        try:
            raw = (
                self._opener.open(request)
                if timeout is None
                else self._opener.open(request, timeout=timeout)
            )
        except urllib.error.HTTPError as error:
            # urllib raises on 4xx/5xx rather than returning them. The error is
            # itself a readable response, and `ApiError` is a better shape to
            # hand a caller than `HTTPError`, so unwrap it here.
            with error:
                return Response(error.code, _lower_headers(error.headers), error.read())
        with raw:
            return Response(raw.status, _lower_headers(raw.headers), raw.read())

    def _resolve_path(
        self, method: str, path: str, path_params: Optional[Mapping[str, Any]]
    ) -> str:
        """Fill ``{param}`` placeholders from the call, falling back to config."""

        def replace(match: "re.Match[str]") -> str:
            name = match.group(1)
            value = None if path_params is None else path_params.get(name)
            if value is None:
                value = self._defaults.get(name)
            if value is None or value == "":
                hint = "."
                if name == SCOPE_PARAM:
                    hint = " — pass {0}=..., or set it when constructing the client.".format(
                        SCOPE_KWARG
                    )
                raise ValueError(
                    'Missing path parameter "{0}" for {1} {2}{3}'.format(
                        name, method.upper(), path, hint
                    )
                )
            return urllib.parse.quote(_to_str(value), safe="")

        return _PLACEHOLDER.sub(replace, path)


def _to_str(value: Any) -> str:
    """Render a scalar the way the API reads it, not the way Python prints it."""
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, str):
        return value
    if value is None:
        return ""
    if isinstance(value, (dict, list, tuple)):
        return json.dumps(value)
    return str(value)


def _encode_query(query: Optional[Mapping[str, Any]]) -> str:
    if not query:
        return ""
    pairs: List[Tuple[str, str]] = []
    for key, value in query.items():
        if value is None:
            continue
        if isinstance(value, (list, tuple)):
            pairs.extend((key, _to_str(item)) for item in value if item is not None)
        else:
            pairs.append((key, _to_str(value)))
    return "?" + urllib.parse.urlencode(pairs) if pairs else ""


def _lower_headers(headers: Any) -> Dict[str, str]:
    if headers is None:
        return {}
    items = headers.items() if hasattr(headers, "items") else headers
    return {str(key).lower(): str(value) for key, value in items}


def _escape_quotes(value: str) -> str:
    return value.replace("\\", "\\\\").replace('"', '\\"')


def _multipart_part(name: str, value: Any) -> bytes:
    filename: Optional[str] = None
    content_type: Optional[str] = None
    if isinstance(value, tuple):
        filename = str(value[0])
        payload = bytes(value[1])
        if len(value) > 2:
            content_type = str(value[2])
    elif isinstance(value, (bytes, bytearray)):
        # A bare buffer still has to be sent as a file part, or the server sees
        # a text field. The field name is the only filename available.
        filename = name
        payload = bytes(value)
    else:
        payload = _to_str(value).encode("utf-8")

    disposition = 'form-data; name="{0}"'.format(_escape_quotes(name))
    if filename is not None:
        disposition += '; filename="{0}"'.format(_escape_quotes(filename))
        content_type = content_type or _DEFAULT_PART_TYPE

    out = bytearray()
    out += b"Content-Disposition: " + disposition.encode("utf-8") + _CRLF
    if content_type is not None:
        out += b"Content-Type: " + content_type.encode("utf-8") + _CRLF
    out += _CRLF + payload + _CRLF
    return bytes(out)


def _encode_multipart(fields: Mapping[str, Any]) -> Tuple[bytes, str]:
    """Serialize ``multipart/form-data`` by hand — stdlib has no encoder."""
    boundary = "----infrawrench{0}".format(uuid.uuid4().hex)
    marker = boundary.encode("ascii")
    out = bytearray()
    for name, value in fields.items():
        if value is None:
            continue
        # A list means the same field repeated; a tuple is one file spec.
        parts = value if isinstance(value, list) else [value]
        for part in parts:
            if part is None:
                continue
            out += b"--" + marker + _CRLF + _multipart_part(name, part)
    out += b"--" + marker + b"--" + _CRLF
    return bytes(out), "multipart/form-data; boundary={0}".format(boundary)


def _to_api_error(response: Response, method: str, url: str) -> ApiError:
    text = response.body.decode("utf-8", "replace") if response.body else ""
    body: Any = text
    if text:
        try:
            body = json.loads(text)
        except ValueError:
            pass  # Not JSON — keep the raw text as the body.
    record = body if isinstance(body, dict) else {}
    detail = record.get("error") or record.get("message")
    if not isinstance(detail, str):
        detail = "{0} Request failed".format(response.status)
    code = record.get("code")
    return ApiError(
        status=response.status,
        message="{0} {1} failed: {2}".format(method, url, detail),
        code=code if isinstance(code, str) else None,
        body=body,
        method=method,
        url=url,
    )

File: app/packages/web/scripts/sdk/targets/typescript/runtime.ts
/**
 * The hand-written half of the generated TypeScript SDK.
 *
 * This file is **not imported by anything in this repo** — the generator reads
 * it off disk and inlines it at the top of `sdk/typescript/index.ts`, ahead of
 * the generated models and namespace classes. Keeping it as real source (rather
 * than a template string) means it is typechecked by `pnpm typecheck` and
 * formatted by `pnpm format` like everything else, so a mistake in the request
 * plumbing shows up here instead of in generated output nobody reads.
 *
 * Two tokens are substituted at generation time — see `SUBSTITUTIONS` in
 * `./index.ts`. They are written as plain string literals so this file still
 * typechecks on its own.
 *
 * Nothing here may import anything: the published package has zero
 * dependencies and must run on stock Node 18+ / any modern browser.
 */

// --8<-- everything below this line is inlined into the generated SDK --8<--

/** Replaced with the first server advertised by the spec. */
const DEFAULT_BASE_URL = "@@BASE_URL@@";

/**
 * Replaced with the path parameter the client can carry as configuration
 * (`orgId`), or `null` if the API has no such parameter.
 */
const SCOPE_PARAM: string | null = "@@SCOPE_PARAM@@";

export interface ClientOptions {
  /** Base URL of the deployment. Defaults to the production API. */
  baseUrl?: string;
  /**
   * API key or WorkOS access token, sent as `Authorization: Bearer <token>`.
   * Omit it only if you are supplying the header yourself.
   */
  apiKey?: string;
  /**
   * Default organization id. Every org-scoped call accepts `orgId`; set it here
   * once and you can leave it off the call sites.
   */
  orgId?: string;
  /** Headers merged into every request. Per-call headers win. */
  headers?: Record<string, string>;
  /** Abort requests after this many milliseconds. No timeout by default. */
  timeoutMs?: number;
  /** Swap the fetch implementation — for tests, proxies, or custom agents. */
  fetch?: typeof globalThis.fetch;
}

/** Per-call overrides, accepted as the last argument of every method. */
export interface RequestOptions {
  /** Cancel the request. Combined with `timeoutMs` if both are set. */
  signal?: AbortSignal;
  /** Headers for this call only. */
  headers?: Record<string, string>;
  /** Timeout for this call only, overriding the client-wide one. */
  timeoutMs?: number;
}

/**
 * Thrown for any non-2xx response. The parsed response body is on `.body`, and
 * `.code` carries the machine-readable discriminator when the API sends one
 * (e.g. `reauthentication_required` on a step-up 403) — branch on that, not on
 * the message.
 */
export class ApiError extends Error {
  readonly status: number;
  readonly code: string | undefined;
  readonly body: unknown;
  readonly method: string;
  readonly url: string;

  constructor(init: {
    status: number;
    message: string;
    code: string | undefined;
    body: unknown;
    method: string;
    url: string;
  }) {
    super(init.message);
    this.name = "ApiError";
    this.status = init.status;
    this.code = init.code;
    this.body = init.body;
    this.method = init.method;
    this.url = init.url;
  }
}

/** A single HTTP call, as described by the generated namespace methods. */
export interface RequestSpec {
  method: string;
  /** URL template with `{param}` placeholders, e.g. `/api/org/{orgId}/accounts`. */
  path: string;
  pathParams?: Record<string, unknown> | undefined;
  query?: Record<string, unknown> | undefined;
  /** JSON request body. */
  body?: unknown;
  /** `multipart/form-data` fields. Mutually exclusive with `body`. */
  form?: Record<string, unknown> | undefined;
  /** What the endpoint returns. Defaults to `json`. */
  accept?: "json" | "binary" | "empty" | undefined;
}

/**
 * Request plumbing shared by every namespace.
 *
 * Exported because the generated namespace classes take one in their
 * constructor; it is not part of the stable surface, and you should reach for
 * `APIV1Client` instead.
 */
export class ApiTransport {
  /** Normalized base URL, without a trailing slash. */
  readonly baseUrl: string;
  private readonly apiKey: string | undefined;
  private readonly defaults: Record<string, string | undefined>;
  private readonly headers: Record<string, string>;
  private readonly timeoutMs: number | undefined;
  private readonly fetchImpl: typeof globalThis.fetch;

  constructor(options: ClientOptions = {}) {
    this.baseUrl = (options.baseUrl ?? DEFAULT_BASE_URL).replace(/\/+$/, "");
    this.apiKey = options.apiKey;
    this.defaults = SCOPE_PARAM === null ? {} : { [SCOPE_PARAM]: options.orgId };
    this.headers = { ...options.headers };
    this.timeoutMs = options.timeoutMs;
    const impl = options.fetch ?? globalThis.fetch;
    if (typeof impl !== "function") {
      throw new TypeError(
        "No global fetch is available in this runtime — pass one as `fetch` in ClientOptions.",
      );
    }
    this.fetchImpl = impl;
  }

  async request<T>(spec: RequestSpec, options: RequestOptions = {}): Promise<T> {
    const url = this.baseUrl + this.resolvePath(spec) + serializeQuery(spec.query);

    const headers: Record<string, string> = {
      accept: spec.accept === "binary" ? "application/octet-stream" : "application/json",
      ...this.headers,
      ...options.headers,
    };
    if (this.apiKey !== undefined && this.apiKey !== "") {
      headers["authorization"] = `Bearer ${this.apiKey}`;
    }

    let body: BodyInit | undefined;
    if (spec.form !== undefined) {
      // Leave content-type unset: fetch has to add the multipart boundary.
      body = toFormData(spec.form);
    } else if (spec.body !== undefined) {
      headers["content-type"] = "application/json";
      body = JSON.stringify(spec.body);
    }

    const signal = combineSignals(options.signal, options.timeoutMs ?? this.timeoutMs);
    // Built up rather than declared inline: `RequestInit` types `body` as
    // `BodyInit | null`, so handing it an explicit `undefined` is a type error.
    const init: RequestInit = { method: spec.method, headers };
    if (body !== undefined) init.body = body;
    if (signal !== undefined) init.signal = signal;

    const response = await this.fetchImpl(url, init);

    if (!response.ok) throw await toApiError(response, spec.method, url);
    if (spec.accept === "binary") return (await response.blob()) as T;
    if (spec.accept === "empty" || response.status === 204 || response.status === 205) {
      return undefined as T;
    }

    const text = await response.text();
    if (text === "") return undefined as T;
    if (!(response.headers.get("content-type") ?? "").includes("json")) return text as T;
    return JSON.parse(text) as T;
  }

  /** Fill `{param}` placeholders from the call, falling back to client config. */
  private resolvePath(spec: RequestSpec): string {
    return spec.path.replace(/\{([^}]+)\}/g, (_match, name: string) => {
      const value = spec.pathParams?.[name] ?? this.defaults[name];
      if (value === undefined || value === null || value === "") {
        const hint =
          name === SCOPE_PARAM
            ? ` — pass it, or set \`${name}\` when constructing the client.`
            : ".";
        throw new TypeError(
          `Missing path parameter "${name}" for ${spec.method} ${spec.path}${hint}`,
        );
      }
      return encodeURIComponent(String(value));
    });
  }
}

function serializeQuery(query: Record<string, unknown> | undefined): string {
  if (query === undefined) return "";
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(query)) {
    if (value === undefined || value === null) continue;
    if (Array.isArray(value)) {
      for (const item of value as unknown[]) {
        if (item !== undefined && item !== null) params.append(key, String(item));
      }
    } else {
      params.append(key, String(value));
    }
  }
  const encoded = params.toString();
  return encoded === "" ? "" : `?${encoded}`;
}

function toFormData(fields: Record<string, unknown>): FormData {
  const form = new FormData();
  for (const [key, value] of Object.entries(fields)) {
    if (value === undefined || value === null) continue;
    if (value instanceof Blob) form.append(key, value);
    else form.append(key, String(value));
  }
  return form;
}

function combineSignals(
  signal: AbortSignal | undefined,
  timeoutMs: number | undefined,
): AbortSignal | undefined {
  if (timeoutMs === undefined) return signal;
  const timeout = AbortSignal.timeout(timeoutMs);
  if (signal === undefined) return timeout;
  // `AbortSignal.any` is Node 20.3+ / modern browsers. Without it the caller's
  // own signal wins, which is the safer of the two to lose.
  const combine = (AbortSignal as unknown as { any?: (signals: AbortSignal[]) => AbortSignal }).any;
  return typeof combine === "function" ? combine.call(AbortSignal, [signal, timeout]) : signal;
}

async function toApiError(response: Response, method: string, url: string): Promise<ApiError> {
  const text = await response.text().catch(() => "");
  let body: unknown = text;
  if (text !== "") {
    try {
      body = JSON.parse(text);
    } catch {
      // Not JSON — keep the raw text as the body.
    }
  }
  const record: Record<string, unknown> =
    typeof body === "object" && body !== null ? (body as Record<string, unknown>) : {};
  const detail =
    typeof record["error"] === "string"
      ? record["error"]
      : typeof record["message"] === "string"
        ? record["message"]
        : `${response.status} ${response.statusText || "Request failed"}`;
  return new ApiError({
    status: response.status,
    message: `${method} ${url} failed: ${detail}`,
    code: typeof record["code"] === "string" ? record["code"] : undefined,
    body,
    method,
    url,
  });
}
